environments/servlet/core/src/main/java/org/jboss/weld/environment/jetty/JettyContainer.java

```diff
@@ -55,24 +57,49 @@ protected String classToCheck() {
 
     @Override
     public boolean touch(ResourceLoader resourceLoader, ContainerContext context) throws Exception {
-        ServletContext sc = context.getServletContext();
-        // The jetty decorate module from 9.4.20 sets this attribute to indicate that a DecoratingListener is registered.
-        return sc.getAttribute(JETTY_DECORATING_ATTRIBUTE) instanceof String ||
-            JETTY_CDI_VALUE.equals(sc.getAttribute(JETTY_CDI_ATTRIBUTE));
+        ServletContext servletContext = context.getServletContext();
+        // The jetty cdi modules from 9.4.20 sets JETTY_CDI_ATTRIBUTE to indicate that a
+        // DecoratingListener is registered. Weld-3.1.2.Final documented that the decorate module
+        // could be used directly, which sets DECORATING_LISTENER_ATTRIBUTE
+        return servletContext.getAttribute(JETTY_CDI_ATTRIBUTE) instanceof String ||
+            servletContext.getAttribute(DECORATING_LISTENER_ATTRIBUTE) instanceof String;
     }
 
     @Override
     public void initialize(ContainerContext context) {
         try {
             ServletContext servletContext = context.getServletContext();
-            // Is the Jetty server doing its own CDI SPI integration?
-            if (JettyContainer.JETTY_CDI_VALUE.equals(servletContext.getAttribute(JettyContainer.JETTY_CDI_ATTRIBUTE))) {
-                // Yes, no further integration required
-                JettyLogger.LOG.jettyCdiSpiIsSupported();
-            } else {
-                // No, we need to initialize a JettyWeldInjector and WeldDecorator for it
-                super.initialize(context);
-                WeldDecorator.process(servletContext);
+            String mode = (String)servletContext.getAttribute(JETTY_CDI_ATTRIBUTE);
+            if (mode == null) {
+                mode = DECORATING_LISTENER_MODE;
+            }
+            switch (mode) {
+                case CDI_SPI_DECORATOR_MODE:
+                    // For use with the cdi-spi module
+                    // No further integration required
+                    JettyLogger.LOG.jettyCdiSpiIsSupported();
+                    break;
+
+                case CDI_DECORATING_LISTENER_MODE:
+                    // For use with the cdi-decorate module
+                    // Initialize a JettyWeldInjector and create WeldDecorator for it
+                    super.initialize(context);
+                    servletContext.setAttribute(CDI_DECORATING_LISTENER_ATTRIBUTE, new WeldDecorator(servletContext));
+                    JettyLogger.LOG.jettyDecorationIsSupported();
+                    break;
+
+                case DECORATING_LISTENER_MODE:
+                    // For use with the decorate module
+                    // This mode is only needed to match the Weld-3.1.2 documentation.
+                    // Initialize a JettyWeldInjector and create WeldDecorator for it
```

Comment:
I am kinda lost in what's the actual difference in `DECORATING_LISTENER_MODE` and `CDI_DECORATING_LISTENER_MODE` and why we need both?
Cause you elaborate a bit please?

Reply:
The `DECORATING_LISTENER_MODE` is not needed other than that is the style of usage that we documented in the Weld-3.1.2 release, so by supporting it we have backwards compatibility with the integration introduced in 3.1.2.  Essentially it is an identical mechanism to `CDI_DECORATING_LISTENER_MODE` but uses the unrenamed attributes.

Going forward `CDI_DECORATING_LISTENER_MODE` is the preferred integration mode.

Reply:
Alright, but I still don't see the backward compat problem - all users do is use Weld servlet dependency and then on Jetty side they activate a module. They don't care what attributes are used underneath.
So you either end up with Weld 3.1.2 and new Jetty - in which case it depends whether Jetty will recognize that.
Or you will have this new attribute in Weld 3.1.3 in which case it works as expected on Jetty side.

I am probably missing some obvious scenario in which it would blow up...

Reply:
In Jetty we now have the raw `DecoratingListener` mechanism that can be instantiated in a number of ways and has a particular default setup.   We also have the `CdiDecoratingListener` mechanism which specializes `DecoratingListener` and changes the defaults to be more suitable for CDI.

In Weld-3.1.2 there were instructions on how to use the raw `DecoratingListener`, which if followed will configure jetty in a specific way that will be detected by the `DECORATING_LISTENER_MODE`.      

With this PR, we change those instructions to configure jetty to use the `CdiDecoratingListener` and that matching the `CDI_DECORATING_LISTENER_MODE`.

So the backward compatibility problem is only that Weld-3.1.2 documentation might be used with later releases.  It's not a big deal, but might save a little bit of confusion.   Perhaps we should deprecate the `DECORATING_LISTENER_MODE` and produce a warning if it is used?

Reply:
Alright, thanks for explanation.